### PR TITLE
Larva queue QOL

### PR DESCRIPTION
**Pull request by BraveMole**
## About The Pull Request

Action added to observers. Also you can leave the queue by clicking on the action another time

## Why It's Good For The Game

QOL

## Changelog
:cl:
qol: You can leave the larva queue. Also, a new action to join/leave the queue gave to observers
/:cl:

### CODE REVIEW

File: code/datums/actions/observer_action.dm [hunk 1]
```diff
@@ -29,3 +29,14 @@
 	if(!can_use_action())
 		return FALSE
 	check_hive_status(usr)
+
+/datum/action/observer_action/join_larva_queue
+	name = "Join larva queue"
```
Comment (XSlayer300) [inline code/datums/actions/observer_action.dm:34]:
```suggestion
	name = "Join Larva Queue"
```
consistency.

File: code/modules/mob/living/carbon/xenomorph/hive_datum.dm [hunk 2]
```diff
@@ -1191,12 +1194,23 @@ to_chat will check for valid clients itself already so no need to double check f
...
 	RegisterSignal(observer, COMSIG_PARENT_QDELETING, .proc/clean_observer)
 	observer.larva_position =  LAZYLEN(candidate)
 	to_chat(observer, "<span class='warning'>There are no burrowed larvas or no silos. You are in position [observer.larva_position] to become a xeno</span>")
+	return TRUE
+
+/// Remove an observer from the larva candidate queue
+/datum/hive_status/proc/remove_from_larva_candidate_queue(mob/dead/observer/observer)
+	LAZYREMOVE(candidate, observer)
+	UnregisterSignal(observer, COMSIG_PARENT_QDELETING)
+	to_chat(observer, "<span class='warning'>You left the larva queue</span>")
```
Comment (XSlayer300) [inline code/modules/mob/living/carbon/xenomorph/hive_datum.dm:1209]:
```suggestion
	to_chat(observer, "<span class='warning'>You left the Larva queue.</span>")
```

File: code/modules/mob/living/carbon/xenomorph/hive_datum.dm [hunk 3]
```diff
@@ -1191,12 +1194,23 @@ to_chat will check for valid clients itself already so no need to double check f
...
 		return
 	if(LAZYFIND(candidate, observer))
-		to_chat(observer, "<span class='warning'>You are already in queue!</span>")
-		return
+		remove_from_larva_candidate_queue(observer)
+		return FALSE
 	LAZYADD(candidate, observer)
 	RegisterSignal(observer, COMSIG_PARENT_QDELETING, .proc/clean_observer)
 	observer.larva_position =  LAZYLEN(candidate)
 	to_chat(observer, "<span class='warning'>There are no burrowed larvas or no silos. You are in position [observer.larva_position] to become a xeno</span>")
```
Comment (XSlayer300) [inline code/modules/mob/living/carbon/xenomorph/hive_datum.dm:1202]:
```suggestion
	to_chat(observer, "<span class='warning'>There are no burrowed Larvae or no silos. You are in position [observer.larva_position] to become a Xenomorph</span>")
```
Maybe an additional fix too.

File: code/modules/mob/living/carbon/xenomorph/hive_datum.dm [hunk 4]
```diff
@@ -1191,12 +1194,23 @@ to_chat will check for valid clients itself already so no need to double check f
...
+	return TRUE
+
+/// Remove an observer from the larva candidate queue
+/datum/hive_status/proc/remove_from_larva_candidate_queue(mob/dead/observer/observer)
+	LAZYREMOVE(candidate, observer)
+	UnregisterSignal(observer, COMSIG_PARENT_QDELETING)
+	to_chat(observer, "<span class='warning'>You left the Larva queue.</span>")
+	var/mob/dead/observer/observer_in_queue
+	for(var/i in 1 to LAZYLEN(candidate))
+		observer_in_queue = LAZYACCESS(candidate, i)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/hive_datum.dm:1212]: why are you doing a lazyaccess when you already check length?

**Review by TiviPlus — CHANGES_REQUESTED** [review]

**Review by TiviPlus — APPROVED** [review]

**Comment by usnpeepoo** [comment]
when will burrowed larva count towards larva queue so you can observe and wait for a burrowed to appear without needing a silo

**Pull request merged**